oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/logout/OidcLogoutSuccessHandler.java

```diff
@@ -0,0 +1,89 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.client.oidc.logout;
+
+import java.nio.charset.StandardCharsets;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.springframework.security.core.Authentication;
+import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken;
+import org.springframework.security.oauth2.client.registration.ClientRegistration;
+import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository;
+import org.springframework.security.oauth2.core.oidc.user.OidcUser;
+import org.springframework.security.web.authentication.logout.SimpleUrlLogoutSuccessHandler;
+import org.springframework.util.Assert;
+import org.springframework.web.util.UriComponentsBuilder;
+
+/**
+ * A logout success handler for coordinating logout with an OIDC OP through the user agent.
+ *
+ * @author Josh Cummings
+ * @since 5.2
+ * @see <a href="http://openid.net/specs/openid-connect-session-1_0.html#RPLogout">RP-Initiated Logout</a>
+ * @see {@link org.springframework.security.web.authentication.logout.LogoutSuccessHandler}
+ */
+public class OidcLogoutSuccessHandler extends SimpleUrlLogoutSuccessHandler {
+	private final ClientRegistrationRepository repository;
+
+	private String postLogoutRedirectUri;
+
+	public OidcLogoutSuccessHandler(ClientRegistrationRepository repository) {
+		Assert.notNull(repository, "repository cannot be null");
+		this.repository = repository;
+	}
+
+	@Override
+	protected String determineTargetUrl(HttpServletRequest request,
+			HttpServletResponse response, Authentication authentication) {
+
+		if (authentication instanceof OAuth2AuthenticationToken) {
+			OAuth2AuthenticationToken token = (OAuth2AuthenticationToken) authentication;
+
+			ClientRegistration registration =
+					this.repository.findByRegistrationId(token.getAuthorizedClientRegistrationId());
+
+			String endpointUri = (String)
+					registration.getProviderDetails().getConfigurationMetadata().get("end_session_endpoint");
+
+			UriComponentsBuilder builder = UriComponentsBuilder.fromHttpUrl(endpointUri);
+
+			if (token.getPrincipal() instanceof OidcUser) {
+				OidcUser user = (OidcUser) token.getPrincipal();
+				builder.queryParam("id_token_hint", user.getIdToken().getTokenValue());
+			}
+
+			if (this.postLogoutRedirectUri != null) {
+				builder.queryParam("post_logout_redirect_uri", this.postLogoutRedirectUri);
+			}
+
+			return builder.encode(StandardCharsets.UTF_8).build().toUriString();
+		}
+
+		return getDefaultTargetUrl();
```

Comment:
I think this should be `super.determineTargetUrl(request, response)` ? Also, I would prefer this to happen in the pre-checks to avoid the nested `if` - I find it more readable that way.